virtwho/virt/rhevm/rhevm.py

```diff
@@ -71,18 +71,73 @@ def __init__(self, logger, config):
         self.password = self.config.password
 
-        self.clusters_url = urlparse.urljoin(self.url, "api/clusters")
-        self.hosts_url = urlparse.urljoin(self.url, "api/hosts")
-        self.vms_url = urlparse.urljoin(self.url, "api/vms")
-
         self.auth = HTTPBasicAuth(self.config.username, self.config.password)
 
+    @property
+    def clusters_url(self):
+        if not hasattr(self, '_clusters_url'):
+            if self.major_version == '4' and 'ovirt-engine' not in self.url:
+                self._clusters_url = urlparse.urljoin(self.url, 'ovirt-engine/api/clusters')
+            else:
+                self._clusters_url = urlparse.urljoin(self.url, 'api/clusters')
+        return self._clusters_url
+
+    @property
+    def hosts_url(self):
+        if not hasattr(self, '_hosts_url'):
+            if self.major_version == '4' and 'ovirt-engine' not in self.url:
+                self._hosts_url = urlparse.urljoin(self.url, 'ovirt-engine/api/hosts')
+            else:
+                self._hosts_url = urlparse.urljoin(self.url, 'api/hosts')
+        return self._hosts_url
+
+    @property
+    def vms_url(self):
+        if not hasattr(self, '_vms_url'):
+            if self.major_version == '4' and 'ovirt-engine' not in self.url:
+                self._vms_url = urlparse.urljoin(self.url, 'ovirt-engine/api/vms')
+            else:
+                self._vms_url = urlparse.urljoin(self.url, 'api/vms')
+        return self._vms_url
+
+    @property
+    def major_version(self):
+        if not hasattr(self, '_major_version'):
+            self.getVersion()
+        return self._major_version
+
+    def getVersion(self):
+        try:
+            response = requests.get(urlparse.urljoin(self.url , 'api'), auth=self.auth, verify=False)
+            if response.status_code == 404 and not 'ovirt-engine' in self.url:
+                response = requests.get(urlparse.urljoin(self.url, 'ovirt-engine/api'), auth=self.auth, verify=False)
+            response.raise_for_status()
+        except requests.RequestException as e:
+            raise virt.VirtError("Unable to connect to RHEV-M server: %s" % str(e))
+
+        try:
+            api = ElementTree.fromstring(response.content)
+        except Exception as e:
+            self.logger.debug("Invalid xml file: %s" % response)
+            raise virt.VirtError("Invalid XML file returned from RHEV-M: %s" % str(e))
+        version = api.find('.//version')
+        if version is not None:
+            major = version.attrib['major']
+            self._major_version = major
+        else:
+            self.logger.info("Could not determine version")
+
     def get(self, url):
         """
         Call RHEV-M server and retrieve what's on given url.
         """
         try:
-            response = requests.get(url, auth=self.auth, verify=False)
+            if self.major_version == '4':
+                headers = dict()
+                headers['Version'] = '3'
```

Comment:
Is there any where we can get this version number when we determine the major version number in (getVersions)?